### ISSUE TITLE
Additional ways to specify camera sensor width 

### ISSUE THREAD

**Issue by john-davies** [openMVG/openMVG#918]
_This issue is raised as a result of discussions on "Incorrect match from camera database #794"_

My current suggestions:

### Add a command line option to specify camera sensor width

For example: --sensor-width "Nikon D3100;23.1"

It should be possible to add several "--sensor-width ..." options to cope with the case where the dataset has multiple camera types

### Add a command line option for an interactive mode

- If this option is set and one or more cameras are not found in the database then the user is prompted to supply the sensor width
- if the option is not set and one or more cameras are not found in the database then an approximate focal length is used ( note that I am not sure how this would be calculated )

### Priority order 

I suggest that the following priority order is used for selecting the sensor width:

1. Command line option "--sensor-width"

2. sensor_width_camera_database.txt file

3. User entered / approximate focal length

**Comment by rperrot** [comment]
Hi @john-davies that's a nice suggestion. 

Just a few remarks : 
- With the sensor-width option, we must find a syntax to provide additional sensors. Any suggestions ? (brackets, comma, ...) ? We must also be sure about how the actual command line can handle it. 
- I don't know if it's a desirable behavior to provide automatically a focal length (does a wrong focal length better than no focal length ?). Maybe an option, set by the user could be desirable (maybe a --default-focal)

I disagree with your order priority. Usually if there's a manual setting, it means that the user want to force a behavior (so point 3 must comme first). 

It should be nice too to provide a web based sensor_width_camera_database.txt (to keep a file with the most updated sensor width) : if the sensor width is not present in the current file, it tries to look for the hosted file.

**Comment by john-davies** [comment]
Sorry, long reply :)

> With the sensor-width option, we must find a syntax to provide additional sensors. 
> Any suggestions ? (brackets, comma, ...) ? We must also be sure about how the 
> actual command line can handle it.

Apologies, I wasn't clear. For additional sensors my intention was that the user could supply multiple "sensor-width" options on the command line, e.g.

`... --sensor-width "Nikon D3100;23.1" --sensor-width "Nikon D3200;23.1" ...`

alternatively we could allow comma separated entries, e.g.

`... --sensor-width "Nikon D3100;23.1","Nikon D3200;23.1" ...`

( I'm not sure if the "," would cause problems with command line parsing, I'll check )

>  I don't know if it's a desirable behavior to provide automatically a focal length 
> (does a wrong focal length better than no focal length ?). Maybe an option, set 
> by the user could be desirable (maybe a --default-focal)

I'm not sure about this. The idea came from a discussion on issue #794 where @pmoulon commented:

_The hard part is if you have a dataset with multiple camera type and so you want to specify various sensor_width. Perhaps we could imagine a interactive mode that ask the sensor_size to the user... or have a default fallback that will initialize a default "approximate" focal length._

A "--default-focal" option would be fine I think

> I disagree with your order priority. Usually if there's a manual setting, it means that the user want to force a behavior (so point 3 must comme first).

Apologies again, I was not clear. Point 3 should read:

3. Program asks the user to enter the sensor width(s) manually, e.g.

`Camera Nikon D3100 not found in database, please enter the sensor width ( in mm ): `

So if the user wants to force a behaviour, e.g. override the database entry, then he can use the "--sensor-width " option which will have priority over the database value.

> It should be nice too to provide a web based sensor_width_camera_database.txt 
> (to keep a file with the most updated sensor width) : if the sensor width is not present 
> in the current file, it tries to look for the hosted file.

Excellent idea. I think that it's possible to link to the latest version of a file in GitHub but I will check further.

**Comment by rperrot** [comment]
@john-davies 

Everything is clear. Do you think we could work together on a PR ?

**Comment by john-davies** [comment]
Yes, sure. I think that all of the changes will be in the "openMVG_main_SfMInit_ImageListing" tool. If it's OK then I will write a final summary of the extra options that we propose and how they will operate?

**Comment by rperrot** [comment]
I'm OK with that.

**Comment by john-davies** [comment]
If you think that the following is OK then I'm happy to try to implement some code changes:

`--sensor-width "Nikon D3100;23.1"`
One or more of these options can be specified. The format of the definition string is the same as that in the current "sensor_width_camera_database.txt" file. Camera models specified here will override any values in the database file.

`--no-remote`
If the sensor width is not found on the command line or in the database then the program will attempt to access the latest version of the database file in GitHub. However if the "--no-remote" option is set then the lookup is not attempted.

`--default-focal <value>`
If the sensor width is not found by any of the previous methods and the user has specified a "--default-focal" option then this value is used for any unknown cameras.

`--no-interactive`
If the camera model is not found by any other method then the user will be prompted to enter the sensor width value on the command line. However if the "--no-interactive" option is set then the user is not prompted and the program will exit with an error message as currently. This would be used where user input is not via the command line, for example, where the program is bundled as part of a GUI.

**Comment by rperrot** [comment]
I think it would be preferable to have inverted settings for remove and interactive. In order to keep retrocompatibility, since theses options are new, they should be triggered by the user and not applied directly. 

So `--no-remote` should be `--remote` and `--no-interactive` should be `--interactive`

@pmoulon Do you agree ?

**Comment by pmoulon** [comment]
I agree with @rperrot It's better to have explicit option for the new behavior.

**Comment by john-davies** [comment]
Sure, that's fine. Thanks.

**Comment by john-davies** [comment]
@pmoulon,  @rperrot 

A small clarification for the "--default-focal" option just to make sure that I've got the right idea. In the code ( main_SfMInit_ImageListing.cpp ) we have the following regarding focal lengths:

      // Create the image entry in the list file
      {
        Datasheet datasheet;
        if ( getInfo( sCamModel, vec_database, datasheet ))
        {
          // The camera model was found in the database so we can compute it's approximated focal length
          const double ccdw = datasheet.sensorSize_;
          focal = std::max ( width, height ) * exifReader->getFocal() / ccdw;
        }

where `sensorSize_` is returned from the DB, `width` & `height` refer to the image and `getFocal` is from the EXIF data. `focal` is then calculated and stored in the "sfm_data.json" file:

    "focal_length": 9377.68400493421,


Which value should the "--default-focal" option replace, is it `focal`?

**Comment by pmoulon** [comment]
`--default_focal` must do the same as `--focal`, it will be the focal length in pixel, so the answer is YES, it will output directly the value in the variable focal.

- I also think to add later on a `--default_focal_ratio` in order to have a default multiplier to compute approximate focal length -> `fpix = default_focal_ratio * max(w, h)`

**Comment by john-davies** [comment]
Thank you.

**Comment by john-davies** [comment]
@pmoulon, @rperrot

Sorry, a further question regarding the --remote option. The URL for the db file on GitHub will look something like:

https://raw.githubusercontent.com/openMVG/openMVG/master/src/openMVG/exif/sensor_width_database/sensor_width_camera_database.txt

The obvious solution to get this file is to use something like libcurl ( https://curl.haxx.se/ ), it's available via the libcurl-dev package I think. However I don't have a Windows or Mac development PC and I'm not sure how to include the equivalent functionality on that platform. ( How do required packages like libpng-dev or libjpeg-dev get implemented on Windows/Mac? )

My initial idea was to use the standard C POSIX library to retrieve the database file from GitHub which would probably have been cross platform compatible with no changes but that won't work because the file link is https and not http :)

**Comment by john-davies** [comment]
I've added `--sensor-width`, `--interactive` & `--default_focal`. As described above I'm reluctant to add the `--remote option` without a Windows/Mac machine to test the build on as I'll need to add extra libraries to support the https file fetch. 

@pmoulon, do you want me to add the `--default_focal_ratio`? If so then can you clarify the priority order for `--focal`, `--default_focal` & `--default_focal_ratio`?
